patchwork-fml/src/main/java/net/minecraftforge/fml/common/ObfuscationReflectionHelper.java

```diff
@@ -0,0 +1,375 @@
+					}
+				}
+
+				break;
+			default:
+				throw new IllegalArgumentException("Someones tampered with enums! Got unexpected type " + domain.name());
+			}
+		}
+
+		// It couldn't be found.
+		return name;
+	}
+
+	/**
+	 * Gets the value a field with the specified index in the given class.
+	 * Note: For performance, use {@link #findField(Class, int)} if you are getting the value more than once.
+	 *
+	 * <p>Throws an exception if the field is not found or the value of the field cannot be gotten.</p>
+	 *
+	 * @param classToAccess The class to find the field on.
+	 * @param instance      The instance of the {@code classToAccess}.
+	 * @param fieldIndex    The index of the field in the {@code classToAccess}.
+	 * @param <T>           The type of the value.
+	 * @param <E>           The type of the {@code classToAccess}.
+	 * @return The value of the field with the specified index in the {@code classToAccess}.
+	 * @throws UnableToAccessFieldException If there was a problem getting the field or the value.
+	 * @deprecated Use {@link #getPrivateValue(Class, Object, String)} because field indices change a lot more often than field names do.
+	 * Forge: TODO Remove in 1.15
+	 */
+	@Deprecated
+	@Nullable
+	public static <T, E> T getPrivateValue(Class<? super E> classToAccess, E instance, int fieldIndex) {
+		try {
+			//noinspection unchecked
+			return (T) findField(classToAccess, fieldIndex).get(instance);
+		} catch (Exception e) {
+			LOGGER.error(REFLECTION, "There was a problem getting field index {} from {}", fieldIndex, classToAccess.getName(), e);
+			throw new UnableToAccessFieldException(e);
+		}
+	}
+
+	/**
+	 * Gets the value a field with the specified name in the given class.
+	 * Note: For performance, use {@link #findField(Class, String)} if you are getting the value more than once.
+	 *
+	 * <p>Throws an exception if the field is not found or the value of the field cannot be gotten.</p>
+	 *
+	 * @param classToAccess The class to find the field on.
+	 * @param instance      The instance of the {@code classToAccess}.
+	 * @param fieldName     The intermediary (unmapped) name of the field to find (e.g. "field_5821").
+	 * @param <T>           The type of the value.
+	 * @param <E>           The type of the {@code classToAccess}.
+	 * @return The value of the field with the specified name in the {@code classToAccess}.
+	 * @throws UnableToAccessFieldException If there was a problem getting the field.
+	 * @throws UnableToAccessFieldException If there was a problem getting the value.
+	 */
+	@Nullable
+	public static <T, E> T getPrivateValue(Class<? super E> classToAccess, E instance, String fieldName) {
+		try {
+			//noinspection unchecked
+			return (T) findField(classToAccess, fieldName).get(instance);
+		} catch (UnableToFindFieldException e) {
+			LOGGER.error(REFLECTION, "Unable to locate field {} ({}) on type {}", fieldName,
+					remapName(INameMappingService.Domain.FIELD, fieldName), classToAccess.getName(), e);
```

Comment:
I don't really like looking up the name here again, even if it's just on the error path...

Reply:
We actually have to do the expensive lookup instead of the fast one, because this only throws when the field isn't found. Not doing the expensive lookup would reduce error readability in the event that someone looked up the right field name in the wrong class.